### ISSUE TITLE
false positive with non-empty-string and sprintf

### ISSUE THREAD

**Issue by ppdeblieck** [vimeo/psalm#5040]
When using sprintf to generate a string that string can't be fed into a function where the argument expects a non-empty-string.
A "Argument 1 of func expects non-empty-string, parent type string provided" is thrown.
See: https://psalm.dev/r/55b6f41128

**Comment by orklah** [comment]
There are some edge cases where this is not true when the template only contains flags or empty chars:
https://3v4l.org/3REnd
- Either because params are all empty-strings
- Or because the non-empty-strings params don't match any flag in the template

Getting the right type 100% would require rewriting the string parser and match every flag with every value.

This would be tedious but this would potentially allow emitting Psalm issue for non-matching types for example in`sprintf('%d', '')`

**Comment by ppdeblieck** [comment]
This means the non-empty-string can't be used at all in methods that can get their arguments by sprintf because it'll always fail. The only solution is setting a more generic 'string' type.

If you want to force non-empty Exception messages like in
`throw new CustomException(sprintf('Unrecognized option "%s"', $name));`
(in Symfony "Exception and error message strings must be concatenated using sprintf"; see https://symfony.com/doc/current/contributing/code/standards.html)

How do we annotate this? Or isn't this possible at all?

**Comment by orklah** [comment]
if you put the sprintf result in a variable, you can just add `/** @var non-empty-string $var */` above, Or even `Assert::notEmpty($var)` or similar if you want to be sure

**Comment by boesing** [comment]
Did I miss something? 

Actually, `non-empty-string` is not equivalent with `!empty`. 
Its just that there is at least one character in there.
https://psalm.dev/r/7f4a837969

I've experienced this as well and thus missing this feature. PHPStan somehow parses the 1st argument to the `sprintf` and knows how many "placeholder" are used:
https://phpstan.org/r/9e4475c2-5a37-4908-a558-a0e785e4ebf2

This might be achievable with some sort of `regex` probably.

I've gathered some LoC which I'd expect to be safely treated as `non-empty-string` based on the informations psalm already has:
https://psalm.dev/r/0e712cceb7

When executing, this works as expected: https://3v4l.org/K288q

**Comment by boesing** [comment]
I'd also say, this could be a psalm error because of the usage of `sprintf` without a placeholder (if we are able to verify that there are placeholders in there).

```php
// Same as above, just without the placeholder
nonEmptyString(sprintf(':'));
```

**Comment by weirdan** [comment]
> PHPStan somehow parses the 1st argument to the sprintf and knows how many "placeholder" are used

See https://github.com/phpstan/phpstan-src/blob/c471c7b050e0929daf432288770de673b394a983/src/Rules/Functions/PrintfParametersRule.php

> I'd be happy to work on this, even tho I still lack experience in AST, e.g.

This can probably done as a plugin. You will have to implement `FunctionReturnTypeProviderInterface` where you have access to the AST and can emit issues with `IssueBuffer::accepts()`

**Comment by orklah** [comment]
Note that since my last comment here, the `non-falsy-string` was introduced to disambiguate the special case of '0' that is not empty(as 0 length char) but will still make empty() return true. This means `non-empty-string` is now describing every string except ''. Some of my example above don't reflect that because I used the empty() function to demonstrate emptiness

**Comment by orklah** [comment]
> This can probably done as a plugin.

Don't you think it belong in the core? I don't expect this to be high maintenance and it won't change much in the far future.

**Comment by weirdan** [comment]
Return providers are basically the same in plugins and core, so it can be integrated easily.

**Comment by boesing** [comment]
Oka, I guess I have a working return type provider. Thanks for the hint!
As I am not too familiar with psalm yet, what would be the proper event to analyze the function arguments? 😇

**Comment by orklah** [comment]
> what would be the proper event to analyze the function arguments

I'd expect AfterFunctionCallAnalysisInterface hook to be the right one for both changing the return type and analysing the params

**Comment by boesing** [comment]
I'm currently working on a plugin:
https://github.com/boesing/psalm-plugin-stringf

I'll see how far I get. Currently looks very promising to verify that the output of `sprintf` will be `non-empty-string`.
I did not start with validating the arguments passed to the function.
So with the `IssueBuffer` I can raise issues, right?

Is it possible for plugins to add configuration options? I am thinking of validating stuff like:
```php
sprintf('%s', true);
```

This actually results in a string `1` but I want to raise an issue because a string was expected (due to the `%s` annotation). Since this is very opinionated, I'd like to opt-in for that check and thus wonder if its possible to validate.

Same goes for:
```
$string = 'foo';
sprintf('%d', $string);
```

This actually will be converted to string `0` (wtf...) and I'd suggest using `%s` instead. However, when a `numeric-string` is passed, that might be fine.

----

There are plenty of features which can be added but most of them should be opt-in tho 😅 
So any feedback would be appreciated.

**Comment by weirdan** [comment]
>  So with the `IssueBuffer` I can raise issues, right?

Correct. If you want to introduce new issue types, you will need to extend `PluginIssue`.

> Is it possible for plugins to add configuration options?

Plugins get passed the content of their respective `<plugin>` entries as `SimpleXMLElement`, which could be arbitrary xml.

**Comment by enumag** [comment]
I'm guessing this is a duplicate?

https://psalm.dev/r/baf10b4c9e

**Comment by boesing** [comment]
@enumag Yah, thats why I've written a plugin for that: https://github.com/boesing/psalm-plugin-stringf

**Comment by enumag** [comment]
It's a very minor thing for me, only encountered it once. So I don't wanna add a separate plugin for that. Any plans to have this fixed in psalm itself?
